split: run set_split_mode() on all storage groups during all_storage_groups_split()

nodep opened the pull request:
`tablet_storage_group_manager::all_storage_groups_split()` calls `set_split_mode()` for each of its storage groups to create split ready compaction groups. It does this by iterating through storage groups using `std::ranges::all_of()` which is not guaranteed to iterate through the entire range, and will stop iterating on the first occurrence of the predicate (`set_split_mode()`) returning false. `set_split_mode()` creates the split compaction groups and returns false if the storage group's main compaction group or merging groups are not empty. This means that in cases where the tablet storage group manager has non-empty storage groups, we could have a situation where split compaction groups are not created for all storage groups.
    
The missing split compaction groups are later created in `tablet_storage_group_manager::split_all_storage_groups()` which also calls `set_split_mode()`, and that is the reason why split completes successfully. The problem is that 
`tablet_storage_group_manager::all_storage_groups_split()` runs under a group0 guard, but
`tablet_storage_group_manager::split_all_storage_groups()` does not. This can cause problems with operations which should exclude with compaction group creation. i.e. DROP TABLE/DROP KEYSPACE
    
Fixes #22431

This is a bugfix and should be back ported to versions with tablets: 6.1 6.2 and 2025.1

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14487/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14487/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_tablets2
#### Failed Tests (2/5761):
- [test_decommission_rack_basic.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14487/testReport/junit/topology_custom/test_tablets2/test_decommission_rack_basic_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_decommission_rack_basic.debug.3)
- [test_decommission_rack_basic.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14487/testReport/junit/topology_custom/test_tablets2/test_decommission_rack_basic_debug_3_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_decommission_rack_basic.debug.3)
#### Build Details:
- Duration: 4 hr 25 min
- Builder: spider5.cloudius-systems.com

bhalevy reviewed (approved):
LGTM.  Nice catch!

scylladb-promoter commented:
#### :red_circle: CI State: ABORTED

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14509/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14509/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_tablets2
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5963/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2844/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13971/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14509/)
#### Build Details:
- Duration: 20 hr
- Builder: i-0556bb8fd057eb4ff (m5ad.8xlarge)

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14536/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14536/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_tablets2
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2855/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5973/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13983/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14536/)
#### Failed Tests (1/46931):
- [test_delete_elements_from_a_set](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14536/testReport/junit/alternator_tests/TesterAlternator/Tests___dtest_with_topology_changes___test_delete_elements_from_a_set) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_elements_from_a_set)
#### Build Details:
- Duration: 10 hr
- Builder: spider8.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14556/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14556/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_tablets2
#### Failed Tests (2/8641):
- [test_decommission_rack_basic.release.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14556/testReport/junit/topology_custom/test_tablets2/test_decommission_rack_basic_release_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_decommission_rack_basic.release.6)
- [test_decommission_rack_basic.release.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14556/testReport/junit/topology_custom/test_tablets2/test_decommission_rack_basic_release_6_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_decommission_rack_basic.release.6)
#### Build Details:
- Duration: 5 hr 57 min
- Builder: spider1.cloudius-systems.com

raphaelsc commented:
I am reviewing this on my Monday

nodep commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14556/) ❌ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14556/) **The following new/updated tests ran 100 times for each mode:** 🔹 topology_custom/test_tablets2
> 
> #### Failed Tests (2/8641):
> * [test_decommission_rack_basic.release.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14556/testReport/junit/topology_custom/test_tablets2/test_decommission_rack_basic_release_6) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_decommission_rack_basic.release.6)
> * [test_decommission_rack_basic.release.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14556/testReport/junit/topology_custom/test_tablets2/test_decommission_rack_basic_release_6_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_decommission_rack_basic.release.6)
> 
> #### Build Details:
> * Duration: 5 hr 57 min
> * Builder: spider1.cloudius-systems.com

#22342

scylladb-promoter commented:
#### :red_circle: CI State: ABORTED

:x: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14580/)
#### Build Details:
- Duration: 16 min
- Builder: spider7.cloudius-systems.com

nodep commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14536/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14536/) **The following new/updated tests ran 100 times for each mode:** 🔹 topology_custom/test_tablets2 ✅ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2855/) ❌ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5973/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13983/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14536/)
> 
> #### Failed Tests (1/46931):
> * [test_delete_elements_from_a_set](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14536/testReport/junit/alternator_tests/TesterAlternator/Tests___dtest_with_topology_changes___test_delete_elements_from_a_set) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_elements_from_a_set)
> 
> #### Build Details:
> * Duration: 10 hr
> * Builder: spider8.cloudius-systems.com

Created issue for this failure: #22396

mykaul commented:
@yaronkaikov - please see above.

scylladb-promoter commented:
#### :red_circle: CI State: ABORTED

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14584/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14584/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_tablets2
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2905/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6024/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14033/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14584/)
#### Build Details:
- Duration: 20 hr
- Builder: i-0a4b933c274852bfe (m5ad.8xlarge)

yaronkaikov commented:
> @yaronkaikov - please see above.

Why a single test takes more then an hour?

mykaul commented:
> > @yaronkaikov - please see above.
> 
> Why a single test takes more then an hour?

We somehow feel there's value in running many tests 300 times each. Some take few minutes. https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14584/testReport/topology_custom/test_tablets2/
test_decommission_rack_after_adding_new_rack.debug for example, takes 2-5 minutes each, runs 100 times...

yaronkaikov commented:
> > > @yaronkaikov - please see above.
> > 
> > 
> > Why a single test takes more then an hour?
> 
> We somehow feel there's value in running many tests 300 times each. Some take few minutes. https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14584/testReport/topology_custom/test_tablets2/ test_decommission_rack_after_adding_new_rack.debug for example, takes 2-5 minutes each, runs 100 times...

Well, this was a request made my developers in our last summit, if we want/need to change this, please open an issue and let's decide what is the value we want or need

mykaul commented:
> > > > @yaronkaikov - please see above.
> > > 
> > > 
> > > Why a single test takes more then an hour?
> > 
> > 
> > We somehow feel there's value in running many tests 300 times each. Some take few minutes. https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14584/testReport/topology_custom/test_tablets2/ test_decommission_rack_after_adding_new_rack.debug for example, takes 2-5 minutes each, runs 100 times...
> 
> Well, this was a request made my developers in our last summit, if we want/need to change this, please open an issue and let's decide what is the value we want or need

I think a follow-up was to limit the runtime.
We need to run it in batches of 10s or whatnot, not 100 at once or so and see overall time.

yaronkaikov commented:
> > > > > @yaronkaikov - please see above.
> > > > 
> > > > 
> > > > Why a single test takes more then an hour?
> > > 
> > > 
> > > We somehow feel there's value in running many tests 300 times each. Some take few minutes. https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14584/testReport/topology_custom/test_tablets2/ test_decommission_rack_after_adding_new_rack.debug for example, takes 2-5 minutes each, runs 100 times...
> > 
> > 
> > Well, this was a request made my developers in our last summit, if we want/need to change this, please open an issue and let's decide what is the value we want or need
> 
> I think a follow-up was to limit the runtime. We need to run it in batches of 10s or whatnot, not 100 at once or so and see overall time.

we are running those tests in a batch of 10, only if passed we continue , up to 100 (meaning 10 x 10)

yaronkaikov commented:
> > > > > > > @yaronkaikov - please see above.
> > > > > > 
> > > > > > 
> > > > > > Why a single test takes more then an hour?
> > > > > 
> > > > > 
> > > > > We somehow feel there's value in running many tests 300 times each. Some take few minutes. https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14584/testReport/topology_custom/test_tablets2/ test_decommission_rack_after_adding_new_rack.debug for example, takes 2-5 minutes each, runs 100 times...
> > > > 
> > > > 
> > > > Well, this was a request made my developers in our last summit, if we want/need to change this, please open an issue and let's decide what is the value we want or need
> > > 
> > > 
> > > I think a follow-up was to limit the runtime. We need to run it in batches of 10s or whatnot, not 100 at once or so and see overall time.
> > 
> > 
> > we are running those tests in a batch of 10, only if passed we continue , up to 100 (meaning 10 x 10)
> 
> Cool, so now let's just add total runtime, and cut it off after X batches, if it passes some time. 1h, if you ask me.

Not sure how you suggest doing that. after one hour you want me to kill the process? so it will fail... Let's discuss it in a dedicate issue

nodep commented:
We are not just running 100 times the test that has been added or changed (in this case `test_drop_keyspace_while_split`), but the entire test suite that the test is in. In this case that means running all tests in `topology_custom/test_tablets2`, of which there are currently 29. And that takes time.
Maybe a solution would be to only run the changed/added test case, and not the entire test suite.

mykaul commented on test/topology_custom/test_tablets2.py:
```diff
@@ -1521,6 +1521,59 @@ async def assert_empty_table():
+    await manager.api.disable_tablet_balancing(servers[0].ip_addr)
+
+    # create a table so that it has at least 2 tablets (and storage groups) per shard
+    await cql.run_async("CREATE KEYSPACE test WITH replication = {'class': 'NetworkTopologyStrategy', 'replication_factor': 1} AND tablets = {'initial': 4};")
```
nit: we are trying to move away from test.test and use unique keyspace name (see https://github.com/scylladb/scylladb/pull/22334 )

bhalevy replied:
https://github.com/scylladb/scylladb/pull/22399 is more comprehensive and will override #22334

nodep commented:
> > We are not just running 100 times the test that has been added or changed (in this case `test_drop_keyspace_while_split`), but the entire test suite that the test is in. In this case that means running all tests in `topology_custom/test_tablets2`, of which there are currently 29. And that takes time. Maybe a solution would be to only run the changed/added test case, and not the entire test suite.
> 
> This is what we are doing, if you look at the logs you will see that we are running this:
> 
> ```
> 15:40:27  + bash -c 'set -o pipefail; ./tools/toolchain/dbuild -e ENABLE_KMS_TEST=1 -e ENABLE_GCP_TEST=1 -e AWS_SECRET_ACCESS_KEY=$AWS_SECRET_ACCESS_KEY -e AWS_ACCESS_KEY_ID=$AWS_ACCESS_KEY_ID -e GCP_USER_1_CREDENTIALS=$GCP_USER_1_CREDENTIALS -v $GCP_USER_1_CREDENTIALS:$GCP_USER_1_CREDENTIALS -e GCP_USER_2_CREDENTIALS=$GCP_USER_2_CREDENTIALS -v $GCP_USER_2_CREDENTIALS:$GCP_USER_2_CREDENTIALS -v `pwd`:/scylladir -v /jenkins/coredumps:/jenkins/coredumps -- ./test.py  --tmpdir=/scylladir/testlog/x86_64 --artifacts_dir_url https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14509/artifact --verbose --repeat 10 "topology_custom/test_tablets2::*" --no-gather-metrics 2>&1 | tee tests_console_output_x86_64.txt'
> 15:40:29  Found 30 tests.
> 15:40:32  ================================================================================
> 15:40:32  [N/TOTAL]   SUITE    MODE   RESULT   TEST
> 15:40:32  ------------------------------------------------------------------------------
> 15:40:32  
> 15:45:56  [1/30]     topology_custom release [ PASS ] topology_custom.test_tablets2.3 317.67s
> ...

I think we have a naming confusion here. I think `topology_custom/test_tablets2` is a **test suite**, which contains **test cases** like `topology_custom/test_tablets2::test_tablet_split` or `topology_custom/test_tablets2::test_concurrent_tablet_repair_and_split`. We are currently running the entire test suite 10x10 times which is visible from the log you quoted:

`--repeat 10 "topology_custom/test_tablets2::*"`

What I am suggesting is to run only the changed/added test case, which should look something  like this:

`--repeat 10 "topology_custom/test_tablets2::test_drop_keyspace_while_split"`

yaronkaikov commented:
> > > We are not just running 100 times the test that has been added or changed (in this case `test_drop_keyspace_while_split`), but the entire test suite that the test is in. In this case that means running all tests in `topology_custom/test_tablets2`, of which there are currently 29. And that takes time. Maybe a solution would be to only run the changed/added test case, and not the entire test suite.
> > 
> > 
> > This is what we are doing, if you look at the logs you will see that we are running this:
> > ```
> > 15:40:27  + bash -c 'set -o pipefail; ./tools/toolchain/dbuild -e ENABLE_KMS_TEST=1 -e ENABLE_GCP_TEST=1 -e AWS_SECRET_ACCESS_KEY=$AWS_SECRET_ACCESS_KEY -e AWS_ACCESS_KEY_ID=$AWS_ACCESS_KEY_ID -e GCP_USER_1_CREDENTIALS=$GCP_USER_1_CREDENTIALS -v $GCP_USER_1_CREDENTIALS:$GCP_USER_1_CREDENTIALS -e GCP_USER_2_CREDENTIALS=$GCP_USER_2_CREDENTIALS -v $GCP_USER_2_CREDENTIALS:$GCP_USER_2_CREDENTIALS -v `pwd`:/scylladir -v /jenkins/coredumps:/jenkins/coredumps -- ./test.py  --tmpdir=/scylladir/testlog/x86_64 --artifacts_dir_url https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14509/artifact --verbose --repeat 10 "topology_custom/test_tablets2::*" --no-gather-metrics 2>&1 | tee tests_console_output_x86_64.txt'
> > 15:40:29  Found 30 tests.
> > 15:40:32  ================================================================================
> > 15:40:32  [N/TOTAL]   SUITE    MODE   RESULT   TEST
> > 15:40:32  ------------------------------------------------------------------------------
> > 15:40:32  
> > 15:45:56  [1/30]     topology_custom release [ PASS ] topology_custom.test_tablets2.3 317.67s
> > ...
> > ```
> 
> I think we have a naming confusion here. I think `topology_custom/test_tablets2` is a **test suite**, which contains **test cases** like `topology_custom/test_tablets2::test_tablet_split` or `topology_custom/test_tablets2::test_concurrent_tablet_repair_and_split`. We are currently running the entire test suite 10x10 times which is visible from the log you quoted:
> 
> `--repeat 10 "topology_custom/test_tablets2::*"`
> 
> What I am suggesting is to run only the changed/added test case, which should look something like this:
> 
> `--repeat 10 "topology_custom/test_tablets2::test_drop_keyspace_while_split"`

Why aren't we showing the specific tests in the log? why only the `suite`?  we use test.py log file to understand what was failing or passing, IMO we should all tests in the logs
> 
> The hard part would be to figure out which test cases have been changed from the PR diff. In this case it is simple, but it can get complicated very quickly.
> 
> But I do agree that some tests are running too long. That could be improved by splitting test suites into smaller ones.

mykaul commented:
> Why aren't we showing the specific tests in the log? why only the `suite`? we use test.py log file to understand what was failing or passing, IMO we should all tests in the logs

Reminds me of https://github.com/scylladb/scylladb/issues/22313

raphaelsc commented:
@nodep I suggest we open a GH issue, since we likely want to backport this.

raphaelsc commented:
@nodep can we still fail the assert if we have a migration in parallel to drop table? schema change is not serialized with migration IIRC, but maybe I remember it wrong.

raphaelsc commented on replica/database.cc:
```diff
@@ -2544,6 +2544,12 @@ future<> database::truncate_table_on_all_shards(sharded<database>& sharded_db, s
     });
 
+    co_await utils::get_local_injector().inject("compaction_disabled_wait", [] (auto& handler) -> future<> {
+        dblog.info("compaction_disabled_wait: wait");
```
I'd rename 'compaction_disabled_wait' to 'truncate_compaction_disabled_wait' to make the scope clearer, and to avoid conflicts in the future.

nodep replied:
I've renamed this.

raphaelsc commented on replica/table.cc:
```diff
@@ -1058,6 +1060,12 @@ sstables::compaction_type_options::split tablet_storage_group_manager::split_com
     sstables::compaction_type_options::split opt = split_compaction_options();
 
+    co_await utils::get_local_injector().inject("compaction_groups_created_wait", [] (auto& handler) -> future<> {
+        dblog.info("compaction_groups_created_wait: waiting");
```
compaction_groups_created_wait -> split_storage_groups_wait

nodep commented:
@yaronkaikov Which builder the tests are run on seems to greatly influence their duration:
The two runs which timed out have run on m5ad.8xlarge machines:

> #### Build Details:
> * Duration: 20 hr
> * Builder: i-0556bb8fd057eb4ff (m5ad.8xlarge)

> #### Build Details:
> * Duration: 20 hr
> * Builder: i-0a4b933c274852bfe (m5ad.8xlarge)

The other runs (which have failed because of flaky tests), were executed on our hardware (as far as I can tell) and seem to have completed a lot quicker:

> #### Build Details:
> * Duration: 4 hr 25 min
> * Builder: spider5.cloudius-systems.com

> #### Build Details:
> * Duration: 10 hr
> * Builder: spider8.cloudius-systems.com

> #### Build Details:
> * Duration: 5 hr 57 min
> * Builder: spider1.cloudius-systems.com

Another thing I have found is that a single `--repeat 10 "topology_custom/test_tablets2::*"` runs about 1h20m on a m5ad.8xlarge:

```
[2025-01-16T13:40:32.045Z] ================================================================================
[2025-01-16T13:40:32.045Z] [N/TOTAL]   SUITE    MODE   RESULT   TEST
[2025-01-16T13:40:32.045Z] ------------------------------------------------------------------------------
[2025-01-16T13:40:32.045Z] 
[2025-01-16T13:45:56.369Z] [1/30]     topology_custom release [ PASS ] topology_custom.test_tablets2.3 317.67s
[2025-01-16T13:45:57.086Z] [2/30]     topology_custom release [ PASS ] topology_custom.test_tablets2.1 324.83s
...
[2025-01-16T14:59:42.302Z] [29/30]    topology_custom  debug  [ PASS ] topology_custom.test_tablets2.9 1357.94s
[2025-01-16T15:01:29.535Z] [30/30]    topology_custom  debug  [ PASS ] topology_custom.test_tablets2.10 1361.74s
[2025-01-16T15:01:29.536Z] ------------------------------------------------------------------------------
[2025-01-16T15:01:29.536Z] CPU utilization: 33.9%
```

but much less (about 30m) on our HW:

```
[2025-01-17T09:54:07.678Z] Found 30 tests.
[2025-01-17T09:54:08.281Z] ================================================================================
[2025-01-17T09:54:08.282Z] [N/TOTAL]   SUITE    MODE   RESULT   TEST
[2025-01-17T09:54:08.282Z] ------------------------------------------------------------------------------
[2025-01-17T09:54:08.282Z] 
[2025-01-17T09:57:56.540Z] [1/30]     topology_custom release [ PASS ] topology_custom.test_tablets2.2 213.19s
[2025-01-17T09:57:56.540Z] [2/30]     topology_custom release [ PASS ] topology_custom.test_tablets2.4 213.88s
...
[2025-01-17T10:24:39.875Z] [29/30]    topology_custom  debug  [ PASS ] topology_custom.test_tablets2.10 555.38s
[2025-01-17T10:24:44.968Z] [30/30]    topology_custom  debug  [ PASS ] topology_custom.test_tablets2.9 559.22s
[2025-01-17T10:24:45.613Z] ------------------------------------------------------------------------------
[2025-01-17T10:24:45.613Z] CPU utilization: 23.7%
```

nodep commented:
> > @nodep can we still fail the assert if we have a migration in parallel to drop table? schema change is not serialized with migration IIRC, but maybe I remember it wrong.
> 
> I think it actually is, since they both go through state machine.

I agree. Schema changes are group0 operations, so they should be serialized with migration.

nodep commented:
New in v3:
- Renamed error injections as suggested
- Moved `test_drop_keyspace_while_split` ftom `test_tablets2.py` to `test_tablets.py` to avoid timeouts on CI (`test_tablets2.py` needs to be split up)

nodep commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14620/) ❌ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14620/) **The following new/updated tests ran 100 times for each mode:** 🔹 topology_custom/test_tablets
> 
> #### Failed Tests (2/1861):
> * [test_replace_with_no_normal_token_owners_in_dc[False].dev.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14620/testReport/junit/topology_custom/test_tablets/test_replace_with_no_normal_token_owners_in_dc_False__dev_5) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_with_no_normal_token_owners_in_dc%5BFalse%5D.dev.5)
> * [test_replace_with_no_normal_token_owners_in_dc[False].dev.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14620/testReport/junit/topology_custom/test_tablets/test_replace_with_no_normal_token_owners_in_dc_False__dev_5_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_with_no_normal_token_owners_in_dc%5BFalse%5D.dev.5)
> 
> #### Build Details:
> * Duration: 1 hr 22 min
> * Builder: spider6.cloudius-systems.com

Already reported CI issue #21905

tgrabiec commented on replica/database.cc:
```diff
@@ -2544,6 +2544,12 @@ future<> database::truncate_table_on_all_shards(sharded<database>& sharded_db, s
         });
     });
 
+    co_await utils::get_local_injector().inject("truncate_compaction_disabled_wait", [] (auto& handler) -> future<> {
```
> This adds a reproducer for https://github.com/scylladb/scylla-enterprise/issues/5100

We shouldn't mention enterprise issues in scylladb.git. For backporting, please open an issue in scylladb.git.

nodep replied:
I've opened a scylladb.git issue and removed enterprise issue reference.

scylladb-promoter commented:
#### :red_circle: CI State: ABORTED

:x: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14623/)
#### Build Details:
- Duration: 13 min
- Builder: spider6.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14625/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14625/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_tablets
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14049/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2919/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6039/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14625/)
#### Failed Tests (2/46571):
- [test_replace_with_no_normal_token_owners_in_dc[False].debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14625/testReport/junit/topology_custom/test_tablets/Tests___Unit_Tests___test_replace_with_no_normal_token_owners_in_dc_False__debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_with_no_normal_token_owners_in_dc[False].debug.1)
- [test_replace_with_no_normal_token_owners_in_dc[False].debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14625/testReport/junit/topology_custom/test_tablets/Tests___Unit_Tests___test_replace_with_no_normal_token_owners_in_dc_False__debug_1_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_with_no_normal_token_owners_in_dc[False].debug.1)
#### Build Details:
- Duration: 8 hr 1 min
- Builder: spider6.cloudius-systems.com

nodep commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14625/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14625/) **The following new/updated tests ran 100 times for each mode:** 🔹 topology_custom/test_tablets ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14049/) ✅ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2919/) ✅ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6039/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14625/)
> 
> #### Failed Tests (2/46571):
> * [test_replace_with_no_normal_token_owners_in_dc[False].debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14625/testReport/junit/topology_custom/test_tablets/Tests___Unit_Tests___test_replace_with_no_normal_token_owners_in_dc_False__debug_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_with_no_normal_token_owners_in_dc%5BFalse%5D.debug.1)
> * [test_replace_with_no_normal_token_owners_in_dc[False].debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14625/testReport/junit/topology_custom/test_tablets/Tests___Unit_Tests___test_replace_with_no_normal_token_owners_in_dc_False__debug_1_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_with_no_normal_token_owners_in_dc%5BFalse%5D.debug.1)
> 
> #### Build Details:
> * Duration: 8 hr 1 min
> * Builder: spider6.cloudius-systems.com

#21905

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14638/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14638/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_tablets
#### Failed Tests (1/5580):
- [test_tablet_rf_change[up].debug.8](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14638/testReport/junit/topology_custom/test_tablets/test_tablet_rf_change_up__debug_8) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_rf_change[up].debug.8)
#### Build Details:
- Duration: 3 hr 9 min
- Builder: spider3.cloudius-systems.com

bhalevy commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14638/) ❌ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14638/) **The following new/updated tests ran 100 times for each mode:** 🔹 topology_custom/test_tablets
> 
> #### Failed Tests (1/5580):
> * [test_tablet_rf_change[up].debug.8](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14638/testReport/junit/topology_custom/test_tablets/test_tablet_rf_change_up__debug_8) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_rf_change%5Bup%5D.debug.8)
> 
> #### Build Details:
> * Duration: 3 hr 9 min
> * Builder: spider3.cloudius-systems.com

```
    @pytest.mark.parametrize("direction", ["up", "down", "none"])
    @pytest.mark.asyncio
    async def test_tablet_rf_change(manager: ManagerClient, direction):
        cfg = {'enable_user_defined_functions': False, 'enable_tablets': True}
        servers = await manager.servers_add(3, config=cfg)
        for s in servers:
            await manager.api.disable_tablet_balancing(s.ip_addr)
    
        cql = manager.get_cql()
        res = await cql.run_async("SELECT data_center FROM system.local")
        this_dc = res[0].data_center
    
        if direction == 'up':
            rf_from = 2
            rf_to = 3
        if direction == 'down':
            rf_from = 3
            rf_to = 2
        if direction == 'none':
            rf_from = 2
            rf_to = 2
    
        await cql.run_async(f"CREATE KEYSPACE test WITH replication = {{'class': 'NetworkTopologyStrategy', '{this_dc}': {rf_from}}}")
        await cql.run_async("CREATE TABLE test.test (pk int PRIMARY KEY, c int);")
        await cql.run_async("CREATE MATERIALIZED VIEW test.test_mv AS SELECT pk FROM test.test WHERE pk IS NOT NULL PRIMARY KEY (pk)")
    
        logger.info("Populating table")
        await asyncio.gather(*[cql.run_async(f"INSERT INTO test.test (pk, c) VALUES ({k}, {k});") for k in range(128)])
    
        async def check_allocated_replica(expected: int):
            replicas = await get_all_tablet_replicas(manager, servers[0], 'test', 'test')
            replicas = replicas + await get_all_tablet_replicas(manager, servers[0], 'test', 'test_mv', is_view=True)
            for r in replicas:
                logger.info(f"{r.replicas}")
                assert len(r.replicas) == expected
    
        logger.info(f"Checking {rf_from} allocated replicas")
        await check_allocated_replica(rf_from)
    
        logger.info(f"Altering RF {rf_from} -> {rf_to}")
>       await cql.run_async(f"ALTER KEYSPACE test WITH replication = {{'class': 'NetworkTopologyStrategy', '{this_dc}': {rf_to}}}")
E       cassandra.OperationTimedOut: errors={'127.135.24.37:9042': 'Client request timeout. See Session.execute[_async](timeout)'}, last_host=127.135.24.37:9042

test/topology_custom/test_tablets.py:153: OperationTimedOut
```

https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14638/artifact/testlog/x86_64/debug/failed_test/test_tablet_rf_change%5Bup%5D.8.debug.8/test_py.log
```

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14687/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14687/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_tablets
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14092/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2958/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6081/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14687/)
#### Failed Tests (4/46584):
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14687/testReport/junit/boost.database_test.test_database_with_data_in_sstables_is_a_mutation_source_reverse_read_back_cg1.1.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [test_rolling_upgrade](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14687/testReport/junit/enterprise_upgrade_test/TestEnterpriseRollingUpgrade/Tests___dtest___test_rolling_upgrade) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_rolling_upgrade)
- [test_rolling_upgrade](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14687/testReport/junit/enterprise_upgrade_test/TestEnterpriseRollingUpgrade/Tests___dtest_with_tablets___test_rolling_upgrade) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_rolling_upgrade)
- [test_rolling_upgrade](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14687/testReport/junit/enterprise_upgrade_test/TestEnterpriseRollingUpgrade/Tests___dtest_with_topology_changes___test_rolling_upgrade) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_rolling_upgrade)
#### Build Details:
- Duration: 8 hr 16 min
- Builder: spider5.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14705/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14705/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_tablets
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2976/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6096/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14108/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14705/)
#### Build Details:
- Duration: 8 hr 24 min
- Builder: spider3.cloudius-systems.com

nodep commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14687/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14687/) **The following new/updated tests ran 100 times for each mode:** 🔹 topology_custom/test_tablets ❌ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14092/) ❌ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2958/) ❌ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6081/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14687/)
> 
> #### Failed Tests (4/46584):
> * [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14687/testReport/junit/boost.database_test.test_database_with_data_in_sstables_is_a_mutation_source_reverse_read_back_cg1.1.xunit/xml/Tests___Unit_Tests____empty_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+%5Bempty%5D)
> * [test_rolling_upgrade](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14687/testReport/junit/enterprise_upgrade_test/TestEnterpriseRollingUpgrade/Tests___dtest___test_rolling_upgrade) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_rolling_upgrade)
> * [test_rolling_upgrade](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14687/testReport/junit/enterprise_upgrade_test/TestEnterpriseRollingUpgrade/Tests___dtest_with_tablets___test_rolling_upgrade) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_rolling_upgrade)
> * [test_rolling_upgrade](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14687/testReport/junit/enterprise_upgrade_test/TestEnterpriseRollingUpgrade/Tests___dtest_with_topology_changes___test_rolling_upgrade) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_rolling_upgrade)
> 
> #### Build Details:
> * Duration: 8 hr 16 min
> * Builder: spider5.cloudius-systems.com

#22315

avikivity commented on test/topology_custom/test_tablets.py:
```diff
@@ -754,3 +754,54 @@ async def test_replace_with_no_normal_token_owners_in_dc(manager: ManagerClient,
+
+    logger.info("Bootstrapping cluster")
+    cmdline = [ '--target-tablet-size-in-bytes', '8192',
+                '--smp', '2' ]
```
Please follow up creating a cluster pool for small tablet tests, so such clusters can be reused.

We should declare the fixures we need up-front, so test.py can schedule the tests for reuse.

I tried to find how long this test takes, but test.py runs all of test_tablets in a single go. @xtrey

avikivity replied:
```console
$ ./test.py --mode dev test_tablets --verbose
Found 8 tests.
================================================================================
[N/TOTAL]   SUITE    MODE   RESULT   TEST
------------------------------------------------------------------------------

[1/8]       cqlpy     dev   [ PASS ] cqlpy.test_tablets.1 3.62s
[2/8]      topology_custom   dev   [ PASS ] topology_custom.test_tablets_cql.1 9.00s
[3/8]      topology_custom   dev   [ PASS ] topology_custom.test_tablets_intranode.1 35.96s
[4/8]      topology_custom   dev   [ PASS ] topology_custom.test_tablets_merge.1 41.75s
[5/8]      topology_custom   dev   [ PASS ] topology_custom.test_tablets.1 216.52s
[6/8]      topology_custom   dev   [ PASS ] topology_custom.test_tablets_migration.1 239.07s
[7/8]      topology_custom   dev   [ PASS ] topology_custom.test_tablets_removenode.1 153.07s
[8/8]      topology_custom   dev   [ PASS ] topology_custom.test_tablets2.1 338.35s
------------------------------------------------------------------------------
CPU utilization: 4.9%
```

xtrey replied:
This is how the test.py works: boost tests executed by function, python tests by file. Cgroups created on test.py level, and pytest knows nothing about that. Potentially, we can add a fixture, so each test function will create its own cgroup as well.
